bson/src/main/org/bson/codecs/jsr310/DateTimeBasedCodec.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs.jsr310;
+
+import org.bson.BsonReader;
+import org.bson.BsonType;
+import org.bson.codecs.Codec;
+import org.bson.codecs.configuration.CodecConfigurationException;
+
+import static java.lang.String.format;
+
+abstract class DateTimeBasedCodec<T> implements Codec<T> {
+
+    long validateAndReadDateTime(final BsonReader reader) {
```

Comment:
Is this check really necessary? Look at what `readDateTime()` does under the hood:
```
    @Override
    public long readDateTime() {
        checkPreconditions("readDateTime", BsonType.DATE_TIME);
        setState(getNextState());
        return doReadDateTime();
    }
...
    protected void checkPreconditions(final String methodName, final BsonType type) {
        if (isClosed()) {
            throw new IllegalStateException("BsonWriter is closed");
        }

        verifyBSONType(methodName, type);
    }
...
    protected void verifyBSONType(final String methodName, final BsonType requiredBsonType) {
        if (state == State.INITIAL || state == State.SCOPE_DOCUMENT || state == State.TYPE) {
            readBsonType();
        }
        if (state == State.NAME) {
            // ignore name
            skipName();
        }
        if (state != State.VALUE) {
            throwInvalidState(methodName, State.VALUE);
        }
        if (currentBsonType != requiredBsonType) {
            throw new BsonInvalidOperationException(format("%s can only be called when CurrentBSONType is %s, "
                                                           + "not when CurrentBSONType is %s.",
                                                           methodName, requiredBsonType, currentBsonType));
        }
    }
```
It is not documented in `BsonReader` interface (an oversight or intentional?), but from what I can see it has only one implementation and it reads a value of the expected type or throws `BsonInvalidOperationException`.

Maybe your intention is to throw `CodecConfigurationException` instead of the above one, but in this case:
* It might be better to simply wrap the internal exception as long as `BsonReader` is not expected to have more implementations not throwing it (which would be strange since many other codecs don't have additional checks, look at `DateCodec` for instance).
* Why is `CodecConfigurationException` better than `BsonInvalidOperationException`? According to the docs the second one makes more sense to me than the first one.

Reply:
I have checked a few other codecs (`IntegerCodec`, `ShortCodec`, `ByteCodec`, ...) and all of them use `BsonInvalidOperationException` when something goes wrong. I'm pretty sure this convention should be followed.

Reply:
Agreed, it lacks consistency, however the prior art here is the misuse of CodecConfigurationExceptions for codecs that map BsonTypes to alternative structures / types, eg: there isn't a direct BsonType, for example the GeoJson Codecs.

Reply:
Added JAVA-2787 to fix in 4.0 will re-review its use here.  Thanks @cbartosiak.

Reply:
If `CodecConfigurationException` is going to be used in codecs in other places than constructors its documentation should be changed probably, because currently it is confusing:

*An exception indicating that a codec registry has been misconfigured in some way, preventing it from providing a codec for the requested class.*

In some cases it is thrown for perfectly configured codecs, which are simply used in a wrong way. And it looks like it was meant to use in codec registries, not codecs. Well, this will be discussed JAVA-2787 ;)